### PR TITLE
Refactor code

### PR DESCRIPTION
**Pull request by ShahierNashaat**
Pull request for review refactoring the OOP School Library project according to SOLID principles using ruby and OOP.

- Create App class.
- Create main class.
- Follow SOLID principles.

### CODE REVIEW

File: app.rb [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+      create_student
+    when '2'
+      create_teacher
+    else
+      puts 'Wrong Input!'
+      return
+    end
+
+    puts 'Person created successfully'
+  end
```
Comment (Mihndim2020) [inline app.rb]: - [ ] Following the `Single responsibility rule`, all requests for input and input validation should placed in classes/methods whose responsibilities are to request or validate input. Please checkout this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) and kindly fix that.

File: app.rb [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    parent_permission = gets.chomp
+
+    unless parent_permission.upcase == 'Y' || parent_permission.upcase == 'N'
+      puts 'Wrong Input!'
+      return
+    end
+
+    parent_permission = parent_permission.upcase == 'Y'
+    @persons.push(Student.new(age: age.to_i, name: name, parent_permission: parent_permission))
+  end
```
Comment (Mihndim2020) [inline app.rb]: - [ ] Following the `Single responsibility rule`, all requests for input and input validation should placed in classes/methods whose responsibilities are to request or validate input. Please checkout this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) and kindly fix that.

File: app.rb [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    print 'Age: '
+    age = gets.chomp
+
+    print 'Name: '
+    name = gets.chomp
+
+    print 'Specialization: '
+    specialization = gets.chomp
+    @persons.push(Teacher.new(specialization: specialization, age: age.to_i, name: name))
+  end
```
Comment (Mihndim2020) [inline app.rb]: - [ ] Following the `Single responsibility rule`, all requests for input and input validation should placed in classes/methods whose responsibilities are to request or validate input. Please checkout this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) and kindly fix that.

File: app.rb [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    print 'Title: '
+    title = gets.chomp
+
+    print 'Author: '
+    author = gets.chomp
+
+    @books.push(Book.new(title, author))
+
+    puts 'Book created successfully'
+  end
```
Comment (Mihndim2020) [inline app.rb]: - [ ] Following the `Single responsibility rule`, all requests for input and input validation should placed in classes/methods whose responsibilities are to request or validate input. Please checkout this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) and kindly fix that.

File: app.rb [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+      return
+    end
+
+    print 'Date: '
+    date = gets.chomp
+
+    @rentals.push(Rental.new(date, @books[book_index.to_i], @persons[person_index.to_i]))
+
+    puts 'Rental created successfully'
+  end
```
Comment (Mihndim2020) [inline app.rb]: - [ ] Following the `Single responsibility rule`, all requests for input and input validation should placed in classes/methods whose responsibilities are to request or validate input. Please checkout this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) and kindly fix that.

File: app.rb [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+  def list_all_rental_by_id
+    print 'ID of person: '
+    person_id = gets.chomp
+
+    puts 'Rentals:'
+    @rentals.each do |rental|
+      if rental.person.id.to_s == person_id
+        puts "Date: #{rental.date}, Book \"#{rental.book.title}\" by #{rental.book.author}"
+      end
```
Comment (Mihndim2020) [inline app.rb]: - [ ] Following the `Single responsibility rule`, all requests for input and input validation should placed in classes/methods whose responsibilities are to request or validate input. Please checkout this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) and kindly fix that.

**Review by Mihndim2020 — CHANGES_REQUESTED** [review]
### Hi @, 

### Good job so far!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Good commit messages,
:heavy_check_mark: Descriptive Readme file,
:heavy_check_mark: Your app is running free of `Linter` errors.

There are some issues that you still need to work on to get this project approved, but you are almost there!
ALMOST-MINION

# Status: Requires Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app.rb (hunk 1 above)
Comment (ShahierNashaat) [reply app.rb]: Hey @Mihndim2020, What do you mean by input validation? because each input has its own validation.

File: app.rb (hunk 2 above)
Comment (ShahierNashaat) [reply app.rb]: Hey @Mihndim2020, What do you mean by input validation? because each input has its own validation.

File: app.rb (hunk 3 above)
Comment (ShahierNashaat) [reply app.rb]: Hey @Mihndim2020, What do you mean by input validation? because each input has its own validation.

File: app.rb (hunk 4 above)
Comment (ShahierNashaat) [reply app.rb]: Hey @Mihndim2020, What do you mean by input validation? because each input has its own validation.

File: app.rb (hunk 5 above)
Comment (ShahierNashaat) [reply app.rb]: Hey @Mihndim2020, What do you mean by input validation? because each input has its own validation.

File: app.rb (hunk 6 above)
Comment (ShahierNashaat) [reply app.rb]: Hey @Mihndim2020, What do you mean by input validation? because each input has its own validation.

Comment (Mihndim2020) [reply app.rb]: Hello @ShahierNashaat, all request for input should be placed in classes/methods whose responsibilities are to request input. Forget about the validation, that was just a generic case. Let me know if it is clear to you. Thank you.

File: app.rb (hunk 5 above)
Comment (Mihndim2020) [reply app.rb]: Hello @ShahierNashaat, all request for input should be placed in classes/methods whose responsibilities are to request input. Forget about the validation, that was just a generic case. Let me know if it is clear to you. Thank you.

File: app.rb (hunk 4 above)
Comment (Mihndim2020) [reply app.rb]: Hello @ShahierNashaat, all request for input should be placed in classes/methods whose responsibilities are to request input. Forget about the validation, that was just a generic case. Let me know if it is clear to you. Thank you.

File: app.rb (hunk 3 above)
Comment (Mihndim2020) [reply app.rb]: Hello @ShahierNashaat, all request for input should be placed in classes/methods whose responsibilities are to request input. Forget about the validation, that was just a generic case. Let me know if it is clear to you. Thank you.

File: app.rb (hunk 2 above)
Comment (Mihndim2020) [reply app.rb]: Hello @ShahierNashaat, all request for input should be placed in classes/methods whose responsibilities are to request input. Forget about the validation, that was just a generic case. Let me know if it is clear to you. Thank you.

File: app.rb (hunk 1 above)
Comment (Mihndim2020) [reply app.rb]: Hello @ShahierNashaat, all request for input should be placed in classes/methods whose responsibilities are to request input. Forget about the validation, that was just a generic case. Let me know if it is clear to you. Thank you.

**Review by dasileker — APPROVED** [review]
## **Hi @ShahierNashaat,** 

## **Congratulations! 🎉 Project Approved ✅**
Your project is complete!  There is nothing else to say other than... it's time to merge it 🚀 
 


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**